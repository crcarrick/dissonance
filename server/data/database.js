import { Model, Sequelize } from 'sequelize';

import { channel } from './channel/channel.model';
import { message } from './message/message.model';
import { server } from './server/server.model';
import { user } from './user/user.model';

export const connectDatabase = () => {
  const sequelize = new Sequelize(
    process.env.DATABASE_NAME,
    process.env.DATABASE_USERNAME,
    process.env.DATABASE_PASSWORD,
    {
      host: 'localhost',
      dialect: 'postgres',
      logging: false,
    }
  );

  const Channel = channel({ sequelize });
  const Message = message({ sequelize });
  const Server = server({ sequelize });
  const User = user({ sequelize });

  class UserServer extends Model {}

  UserServer.init({}, { sequelize, tableName: 'users_servers' });

  const models = { Channel, Message, Server, User, UserServer };

  Object.keys(models).forEach((key) => {
    if ('associate' in models[key]) {
      models[key].associate(models);
    }
  });

  return { models, sequelize };
};
