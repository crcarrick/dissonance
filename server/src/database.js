export const createDbClient = (knex) =>
  knex(require('./../knexfile')[process.env.NODE_ENV]);
