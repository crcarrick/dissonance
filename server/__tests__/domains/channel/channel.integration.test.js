import { gql } from 'apollo-server';

import { createDbClient } from '@dissonance/database';

import {
  createTestClient,
  generateData,
  migrate,
  rollback,
  seedDatabase,
} from '@dissonance/test-utils';

const knex = jest.requireActual('knex');

describe('Channel Integration', () => {
  const mockData = generateData();
  const { users, channels, servers } = mockData;

  // All Hooks
  let dbClient;
  beforeAll((done) => {
    dbClient = createDbClient(knex);

    migrate(dbClient).then(() => done());
  });

  afterAll((done) => {
    rollback()
      .then(() => dbClient.destroy())
      .then(() => done());
  });

  // Each Hooks
  let gqlClient;
  beforeEach(async (done) => {
    await seedDatabase(dbClient);

    gqlClient = createTestClient({
      context: {
        user: users[0],
      },
      dbClient,
    });

    done();
  });

  test('gets a channel', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetChannel($input: GetChannelInput!) {
          channel(input: $input) {
            id
            name
            serverId
          }
        }
      `,
      variables: {
        input: { id: channels[0].id },
      },
    });

    expect(response.data.channel).toMatchSnapshot();
  });

  test('gets a channels messages', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetChannelWithMessages($input: GetChannelInput!) {
          channel(input: $input) {
            messages {
              id
              text
            }
          }
        }
      `,
      variables: {
        input: { id: channels[0].id },
      },
    });

    expect(response.data.channel).toMatchSnapshot();
  });

  test('gets list of channels', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetChannels {
          channels {
            id
            name
            serverId
          }
        }
      `,
    });

    expect(response.data.channels).toMatchSnapshot();
  });

  test('gets list of channels with messages', async () => {
    const response = await gqlClient.query({
      query: gql`
        query GetChannelsWithMessages {
          channels {
            messages {
              id
              text
            }
          }
        }
      `,
    });

    expect(response.data.channels).toMatchSnapshot();
  });

  test('creates a channel', async () => {
    const response = await gqlClient.mutate({
      mutation: gql`
        mutation CreateChannel($input: CreateChannelInput!) {
          createChannel(input: $input) {
            id
            name
            serverId
          }
        }
      `,
      variables: {
        input: {
          name: 'Testing 123',
          serverId: servers[0].id,
        },
      },
    });

    expect(response.data.createChannel).toMatchSnapshot({
      id: expect.any(String),
    });
  });
});
