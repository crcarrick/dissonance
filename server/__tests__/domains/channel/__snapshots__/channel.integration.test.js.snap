// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel Integration creates a channel 1`] = `
Object {
  "id": Any<String>,
  "name": "Testing 123",
  "serverId": "029beee1-5599-4c4b-880d-c238b3ab3d16",
}
`;

exports[`Channel Integration gets a channel 1`] = `
Object {
  "id": "75d3d6ae-379f-4d09-8b50-0fc37ba721f8",
  "name": "welcome",
  "serverId": "029beee1-5599-4c4b-880d-c238b3ab3d16",
}
`;

exports[`Channel Integration gets a channels messages 1`] = `
Object {
  "messages": Array [
    Object {
      "id": "8779f557-01ee-49e7-84e9-2ab193a1a964",
      "text": "Eos quia aut excepturi numquam est accusamus voluptatem qui.",
    },
  ],
}
`;

exports[`Channel Integration gets list of channels 1`] = `
Array [
  Object {
    "id": "75d3d6ae-379f-4d09-8b50-0fc37ba721f8",
    "name": "welcome",
    "serverId": "029beee1-5599-4c4b-880d-c238b3ab3d16",
  },
  Object {
    "id": "6909a725-9df3-44fd-8cd9-c94072994670",
    "name": "general",
    "serverId": "029beee1-5599-4c4b-880d-c238b3ab3d16",
  },
  Object {
    "id": "5721cc81-f093-48ae-be45-dfea9f8557bb",
    "name": "welcome",
    "serverId": "1267b1ac-3833-420f-b8a4-a4f25dbbc73f",
  },
  Object {
    "id": "1617f2f7-b141-4620-86c5-bdbf044d268f",
    "name": "general",
    "serverId": "1267b1ac-3833-420f-b8a4-a4f25dbbc73f",
  },
  Object {
    "id": "5fd6a8d0-e7b4-4aa6-a9a4-3198ecfa88dd",
    "name": "welcome",
    "serverId": "df5bb0e8-1143-459f-8fa2-cf5034a55764",
  },
  Object {
    "id": "b019461f-52e7-4cd2-b7c1-57082e2115fe",
    "name": "general",
    "serverId": "df5bb0e8-1143-459f-8fa2-cf5034a55764",
  },
]
`;

exports[`Channel Integration gets list of channels with messages 1`] = `
Array [
  Object {
    "messages": Array [
      Object {
        "id": "8779f557-01ee-49e7-84e9-2ab193a1a964",
        "text": "Eos quia aut excepturi numquam est accusamus voluptatem qui.",
      },
    ],
  },
  Object {
    "messages": Array [
      Object {
        "id": "7dfb9998-d327-434a-a18b-afb1a15aae3c",
        "text": "Omnis quos aspernatur soluta quia eaque quas officia.",
      },
    ],
  },
  Object {
    "messages": Array [
      Object {
        "id": "b9e5328c-980b-452a-9902-460dc7914afa",
        "text": "Voluptatem quis sunt assumenda repellendus pariatur odio ut.",
      },
    ],
  },
  Object {
    "messages": Array [
      Object {
        "id": "5d3f39ba-2f94-4e3c-a622-4c7b95a2bac1",
        "text": "Nihil incidunt amet doloribus voluptas nihil.",
      },
    ],
  },
  Object {
    "messages": Array [
      Object {
        "id": "a6bf3774-384a-487d-8f37-7e66e7a5d177",
        "text": "Minus aspernatur corporis.",
      },
    ],
  },
  Object {
    "messages": Array [
      Object {
        "id": "51bc78c4-3ea5-4759-853e-c4c06cea97ca",
        "text": "Tempora et odio est impedit eligendi omnis natus harum.",
      },
    ],
  },
]
`;
